From 20c8fd0b5f687f3ddd29b949542b50dc5fa0896c Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Thu, 7 Aug 2025 23:22:52 -0400
Subject: [PATCH] sechost: Add stubs for ControlServiceExA/W

(cherry picked from commit 36eaf0277c8fd7cf3390e4b84c8c0db2ff536e2a)
---
 dlls/advapi32/advapi32.spec |  4 ++--
 dlls/sechost/sechost.spec   |  4 ++--
 dlls/sechost/service.c      | 20 ++++++++++++++++++++
 3 files changed, 24 insertions(+), 4 deletions(-)

diff --git a/dlls/advapi32/advapi32.spec b/dlls/advapi32/advapi32.spec
index 4a9db829f67..91d389cc309 100644
--- a/dlls/advapi32/advapi32.spec
+++ b/dlls/advapi32/advapi32.spec
@@ -106,8 +106,8 @@
 @ stdcall CommandLineFromMsiDescriptor(wstr ptr ptr)
 # @ stub ComputeAccessTokenFromCodeAuthzLevel
 @ stdcall -import ControlService(long long ptr)
-# @ stub ControlServiceExA
-# @ stub ControlServiceExW
+@ stdcall -import ControlServiceExA(long long long ptr)
+@ stdcall -import ControlServiceExW(long long long ptr)
 @ stdcall -import ControlTraceA(int64 str ptr long)
 @ stdcall -import ControlTraceW(int64 wstr ptr long)
 # @ stub ConvertAccessToSecurityDescriptorA
diff --git a/dlls/sechost/sechost.spec b/dlls/sechost/sechost.spec
index f0ab06d2d6e..7c43341a370 100644
--- a/dlls/sechost/sechost.spec
+++ b/dlls/sechost/sechost.spec
@@ -26,8 +26,8 @@
 @ stdcall CloseServiceHandle(long)
 @ stdcall CloseTrace(int64)
 @ stdcall ControlService(long long ptr)
-@ stub ControlServiceExA
-@ stub ControlServiceExW
+@ stdcall ControlServiceExA(long long long ptr)
+@ stdcall ControlServiceExW(long long long ptr)
 @ stdcall ControlTraceA(int64 str ptr long)
 @ stdcall ControlTraceW(int64 wstr ptr long)
 @ stub ConvertSDToStringSDRootDomainW
diff --git a/dlls/sechost/service.c b/dlls/sechost/service.c
index 320a28d2a91..dc4632e6252 100644
--- a/dlls/sechost/service.c
+++ b/dlls/sechost/service.c
@@ -1097,6 +1097,26 @@ BOOL WINAPI DECLSPEC_HOTPATCH ControlService( SC_HANDLE service, DWORD control,
     return set_error( err );
 }
 
+/******************************************************************************
+ *     ControlServiceExA   (sechost.@)
+ */
+BOOL WINAPI DECLSPEC_HOTPATCH ControlServiceExA( SC_HANDLE service, DWORD control, DWORD level, void *params)
+{
+    FIXME( "%p %ld %ld %p stub!\n", service, control, level, params );
+
+    return set_error( ERROR_CALL_NOT_IMPLEMENTED );
+}
+
+/******************************************************************************
+ *     ControlServiceExW   (sechost.@)
+ */
+BOOL WINAPI DECLSPEC_HOTPATCH ControlServiceExW( SC_HANDLE service, DWORD control, DWORD level, void *params)
+{
+    FIXME( "%p %ld %ld %p stub!\n", service, control, level, params );
+
+    return set_error( ERROR_CALL_NOT_IMPLEMENTED );
+}
+
 /******************************************************************************
  *     QueryServiceStatus   (sechost.@)
  */
